File: sort-cpp/src/testing/Stopwatch.cpp
//
// Created by Jeroen
//

#include "Stopwatch.h"

#include <iostream>
#include <chrono>

void Stopwatch::start() {
    if (running) {
        std::cerr << "Stopwatch is already running!" << std::endl;
        return;
    }
    running = true;
    started = std::chrono::high_resolution_clock::now();
}

void Stopwatch::stop() {
    if (!running) {
        std::cerr << "Stopwatch isn't running!" << std::endl;
        return;
    }
    ended = std::chrono::high_resolution_clock::now();
    running = false;
}

long Stopwatch::getDifference() {
    if (running) {
        std::cerr << "Stopwatch is still running!\nCall stop first!" << std::endl;
        return -1;
    }
    if (started.time_since_epoch().count() < 0 || ended.time_since_epoch().count() < 0) {
        std::cerr << "Stopwatch has not been started!\nNo difference to calculate" << std::endl;
        return -1;
    }
    return std::chrono::duration_cast<std::chrono::nanoseconds>(ended - started).count();
}

File: sort-java/src/main/java/com/jeroen/sort/testing/Stopwatch.java
package com.jeroen.sort.testing;

public class Stopwatch {

    void start() {
        if (running) {
            System.err.println("Stopwatch is already running!");
            return;
        }
        running = true;
        started = System.nanoTime();
    }

    void stop() {
        if (!running) {
            System.err.println("Stopwatch isn't running!");
            return;
        }
        ended = System.nanoTime();
        running = false;
    }

    long getDifference() {
        if (running) {
            System.err.println("Stopwatch is still running!\nCall stop first!");
            return -1;
        }
        if (started < 0 || ended < 0) {
            System.err.println("Stopwatch has not been started!\nNo difference to calculate");
            return -1;
        }
        return ended - started;
    }

    private long started = -1;
    private long ended = -1;
    private boolean running = false;

}
